### ISSUE TITLE
Pass scope configuration to Tarbela from the outside

### ISSUE THREAD

**Issue by ePaul** [zalando/tarbela#30]
**Current situation**

Tarbela's configuration is done in [application.yml](https://github.com/zalando/tarbela/blob/master/src/main/resources/config/application.yml), using environment variables for URIs/paths.

But the needed scopes for both Nakadi and the event producer are still hardcoded to the values used in the prototype (the wholesale warehouse service).

**Acceptance criteria**

All the scope names can be passed from the outside, so that Tarbela can be used in several deployments without any code change for several event producers (and possibly also several Nakadis with different scope names).

**Out of scope**

* Allow having several event producers with the same Tarbela instance. → #34.
* Allow having several event sinks with the same Tarbela instance → #35.

**Comment by ePaul** [comment]
**Analysis**

Currently the token configuration is done by the libraries [`spring-boot-zalando-stups-tokens`](https://github.com/zalando-stups/spring-boot-zalando-stups-tokens).

This seems to need the scope configuration as list items in a `tokens.token-configuration-list` configuration (e.g. in application.yml). This looks like it wouldn't be easily mappable to environment variables.

Maybe we can access (and configure) this token-configuration-list from code, like [mentioned here](https://github.com/zalando-stups/spring-boot-zalando-stups-tokens/issues/3#issuecomment-217014681), and then pass the lists of scopes for both producer and sink in environment variables.

For #34 + #35 we would need a more flexible way of doing that, though, passing a list of configuration objects.

**Comment by ePaul** [comment]
It looks like the YAML configuration can also be passed via environment variables (or command line parameters).

Original YAML:
```yaml
tokens:
    accessTokenUri: https://token.services.auth.zalando.com/oauth2/access_token?realm=/services
    tokenInfoUri: https://info.services.auth.zalando.com/oauth2/tokeninfo
  ### TODO!!!! change this
    credentialsDirectory: ${user.home}/.tarbela-lumberjack-staging/meta/credentials
    token-configuration-list:
        - tokenId: producer
          scopes:
              - uid
              - warehouse-allocation.read
        - tokenId: zalando-nakadi
          scopes: nakadi.event_stream.write

producer.events.uri: https://warehouse-allocation-staging.wholesale.zalan.do/events
nakadi.submission.uriTemplate: https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events
```

Environment variables settings for the call:
```bash
ACCESS_TOKEN_URI='https://token.services.auth.zalando.com/oauth2/access_token?realm=/services' \
TOKEN_INFO_URI='https://info.services.auth.zalando.com/oauth2/tokeninfo' \
CREDENTIALS_DIR='~/.tarbela-lumberjack-staging/meta/credentials/' \
NAKADI_SUBMISSION_URI_TEMPLATE='https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events' \
PRODUCER_EVENTS_URI='https://warehouse-allocation-staging.wholesale.zalan.do/events' \
tokens.token-configuration-list[0].tokenId='producer' \
tokens.token-configuration-list[0].scopes[0]='uid' \
tokens.token-configuration-list[0].scopes[1]='warehouse-allocation.read' \
tokens.token-configuration-list[0].scopes[2]='warehouse-allocation.event_log_write' \
tokens.token-configuration-list[1].tokenId='zalando-nakadi' \
tokens.token-configuration-list[1].scopes[0]='nakadi.event_stream.write' \
java -jar target/tarbela.jar 
```

**Comment by ePaul** [comment]
While this works locally, it doesn't seem to work with Docker + Taupage. I have to figure out why.

**Comment by ePaul** [comment]
Docker's `-e` command line option doesn't seem to properly transfer variables with `.` in the name to the process being run. This doesn't work:

```bash
sudo docker run -it -v /home/paulo/.tarbela-lumberjack-staging/meta/credentials/:/meta/credentials \
    -e tokens.accessTokenUri='https://token.services.auth.zalando.com/oauth2/access_token?realm=/services' \
    -e tokens.tokenInfoUri='https://info.services.auth.zalando.com/oauth2/tokeninfo' \
    -e tokens.credentialsDirectory='/meta/credentials/' \
    -e tokens.token-configuration-list[0].tokenId='producer' \
    -e tokens.token-configuration-list[0].scopes[0]='uid' \
    -e tokens.token-configuration-list[0].scopes[1]='warehouse-allocation.read' \
    -e tokens.token-configuration-list[0].scopes[2]='warehouse-allocation.event_log_write' \
    -e tokens.token-configuration-list[1].tokenId='zalando-nakadi' \
    -e tokens.token-configuration-list[1].scopes[0]='nakadi.event_stream.write' \
    -e nakadi.submission.uriTemplate='https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events' \
    -e 'producer.events.uri=https://warehouse-allocation-staging.wholesale.zalan.do/events' pierone.stups.zalan.do/lumberjack/tarbela-lumberjack:v1o0o1
```

Passing the properties as command line arguments works (but I don't see how I can specify such arguments in Taupage):

```bash
sudo docker run -it -v /home/paulo/.tarbela-lumberjack-staging/meta/credentials/:/meta/credentials \
    pierone.stups.zalan.do/lumberjack/tarbela-lumberjack:v1o0o1 \
    java -jar /tarbela.jar \
   --tokens.accessTokenUri='https://token.services.auth.zalando.com/oauth2/access_token?realm=/services' \
   --tokens.tokenInfoUri='https://info.services.auth.zalando.com/oauth2/tokeninfo' \
   --tokens.credentialsDirectory='/meta/credentials/' \
   --tokens.token-configuration-list[0].tokenId='producer' \
   --tokens.token-configuration-list[0].scopes[0]='uid' \
   --tokens.token-configuration-list[0].scopes[1]='warehouse-allocation.read' \
   --tokens.token-configuration-list[0].scopes[2]='warehouse-allocation.event_log_write' \
   --tokens.token-configuration-list[1].tokenId='zalando-nakadi' \
   --tokens.token-configuration-list[1].scopes[0]='nakadi.event_stream.write' \
   --nakadi.submission.uriTemplate='https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events' \
   --'producer.events.uri=https://warehouse-allocation-staging.wholesale.zalan.do/events'
```

**Comment by ePaul** [comment]
Aha, it looks like Spring (or Spring boot?) has a workaround built in for that problem. When looking for a property, it is also looking at the environment variable with a `_` instead of the `.`.

```bash
sudo docker run -it -v /home/paulo/.tarbela-lumberjack-staging/meta/credentials/:/meta/credentials \
    -e tokens_accessTokenUri='https://token.services.auth.zalando.com/oauth2/access_token?realm=/services' \
    -e tokens_tokenInfoUri='https://info.services.auth.zalando.com/oauth2/tokeninfo' \
    -e tokens_credentialsDirectory='/meta/credentials/' \
    -e tokens_token-configuration-list[0]_tokenId='producer' \
    -e tokens_token-configuration-list[0]_scopes[0]='uid' \
    -e tokens_token-configuration-list[0]_scopes[1]='warehouse-allocation.read' \
    -e tokens_token-configuration-list[0]_scopes[2]='warehouse-allocation.event_log_write' \
    -e tokens_token-configuration-list[1]_tokenId='zalando-nakadi' \
    -e tokens_token-configuration-list[1]_scopes[0]='nakadi.event_stream.write' \
    -e nakadi_submission_uriTemplate='https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events' \
    -e 'producer_events_uri=https://warehouse-allocation-staging.wholesale.zalan.do/events' pierone.stups.zalan.do/lumberjack/tarbela-lumberjack:v1o0o1
```

This works locally. Now trying with Taupage ...

**Comment by ePaul** [comment]
Writing this in TaupageConfig works:

```yaml
environment:
      'tokens_accessTokenUri': https://token.services.auth.zalando.com/oauth2/access_token?realm=/services
      'tokens_tokenInfoUri': https://info.services.auth.zalando.com/oauth2/tokeninfo
      'tokens_credentialsDirectory': /meta/credentials
      'tokens_token-configuration-list[0]_tokenId': producer
      'tokens_token-configuration-list[0]_scopes[0]': uid
      'tokens_token-configuration-list[0]_scopes[1]': warehouse-allocation.read
      'tokens_token-configuration-list[0]_scopes[2]': warehouse-allocation.event_log_write
      'tokens_token-configuration-list[1]_tokenId': zalando-nakadi
      'tokens_token-configuration-list[1]_scopes[0]': nakadi.event_stream.write
      'producer_events_uri': https://warehouse-allocation-staging.wholesale.zalan.do/events,
      'nakadi_submission_uriTemplate': https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events
```

After a hint from @Jan-M I tried the `SPRING_APPLICATION_JSON` environment variable, which allows passing the configuration similarly to the YAML, but in JSON format. Contrary to popular believe, environment variables don't have to be one-line strings, so this actually works (and is quite readable, although it has some syntactic baggage):

        environment:
          SPRING_APPLICATION_JSON: |
             {
               "tokens": {
                 "accessTokenUri": "https://token.services.auth.zalando.com/oauth2/access_token?realm=/services",
                 "tokenInfoUri": "https://info.services.auth.zalando.com/oauth2/tokeninfo",
                 "credentialsDirectory": "/meta/credentials",
                 "token-configuration-list": [
                   {
                     "tokenId": "producer",
                     "scopes": [
                        "uid",
                        "warehouse-allocation.read",
                        "warehouse-allocation.event_log_write"
                     ]
                   },
                   {
                     "tokenId": "zalando-nakadi",
                     "scopes": [
                        "nakadi.event_stream.write"
                     ]
                   }
                 ]
               },
               "producer.events.uri": "https://warehouse-allocation-staging.wholesale.zalan.do/events",
               "nakadi.submission.uriTemplate": "https://nakadi-sandbox.aruha-test.zalan.do/event-types/{type}/events"
             }

**Comment by ePaul** [comment]
Fixed with #30, is in Release v1.1.0.

**Comment by goekhanak** [comment]
tested locally also and on AWS
